capabilities/istio.ts

```diff
@@ -1,15 +1,115 @@
   name: "istio",
   description: "Istio service mesh capability.",
   namespaces: [],
 });
 
-// Use the 'When' function to create a new Capability Action
 const { When } = Istio;
 
-// When(a.<Kind>).Is<Event>().Then(change => change.<changes>
-When(a.Service)
-  .IsCreated()
-  .Then(svc => svc.SetLabel("pepr", "hi-istio"));
+export class IstioVirtualService extends a.GenericKind {
+  spec: VirtualService["spec"];
+  status?: VirtualService["status"];
+  static apiVersion: VirtualService["apiVersion"];
+  static kind: VirtualService["kind"];
+}
+
+RegisterKind(IstioVirtualService, {
+  group: VirtualService.apiVersion.split("/")[0],
+  version: VirtualService.apiVersion.split("/")[1],
+  kind: VirtualService.kind,
+});
+
+// This capability will only create a gateway if:
+// 1. the ingress class's format for namespace is not set (or set to the same namespace as the ingress)
+// 2. the ingress class must not have wildcards in it. This will be SNI and needs to be a complete match
+//
+// If the ingress class's hostname does have a wildcard in it, and the namespace is the same as the ingress, this is an error
+function parseIngressClass(
+  mynamespace: string,
+  input: string
+): { createGateway: boolean; gateway: string } {
+  const regex = /^pepr-(?:(?<namespace>.+?)\.)?(?<gatewayName>.+)$/;
+  const match = input.match(regex);
+
+  if (match) {
+    const { namespace, gatewayName } = match.groups;
+    if (namespace === undefined || namespace === mynamespace) {
+      return {
+        createGateway: true,
+        gateway: gatewayName,
+      };
+    } else {
+      return {
+        createGateway: false,
+        gateway: `${namespace}/${gatewayName}`,
+      };
+    }
+  }
+
+  return {
+    createGateway: false,
+    gateway: undefined,
+  };
+}
+
+// TODO: do we create a custom resource for pepr for the "state store"
+
+When(a.Ingress)
+  .IsCreatedOrUpdated()
+  .Then(async ing => {
+    // if this object is set to deletion, ignore this
+    if (ing.Raw.metadata?.deletionTimestamp !== undefined) {
+      return;
+    }
+    const ingressClass = parseIngressClass(
+      ing.Raw.metadata.namespace,
+      ing.Raw.spec.ingressClassName
+    );
+    if (ingressClass.gateway === undefined) {
+      return;
+    }
+    const k8s = new K8sAPI();
+    await k8s.labelNamespace(ing.Raw.metadata.namespace, {
+      "istio-injection": "enabled",
+    });
+    // TODO: if at this point in the deployment there are deployments and statefulsets created, they might need to be updated
+    try {
+      if (ingressClass.createGateway) {
+        await k8s.upsertGateway(ing.Raw, ingressClass.gateway);
+        ing.SetAnnotation("pepr.dev/istio-gateway", ingressClass.gateway);
+      }
+      await k8s.upsertVirtualService(ing.Raw, ingressClass.gateway);
+      ing.SetAnnotation("pepr.dev/istio-virtualservice", ing.Raw.metadata.name);
+    } catch (e) {
+      console.error("Failed to create or update VirtualService:", e);
```

Comment:
Do we have a plan yet for how Pepr will handle errors? They never really get retried because things only happen once at admission, right?

Reply:
this flow control is going to get a deeper look after Jeff's latest pepr changes go in.